src/Symfony/Bundle/FrameworkBundle/Secret/SodiumEncoder.php

```diff
@@ -0,0 +1,128 @@
+            throw new \LogicException(sprintf('A key already exist in "%s".', $path));
+        }
+
+        $this->encryptionKey = null;
+
+        $encryptionKey = \sodium_crypto_stream_keygen();
+        (new Filesystem())->dumpFile($path, $encryptionKey);
+        \sodium_memzero($encryptionKey);
+
+        return [$path];
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function encrypt(string $message): string
+    {
+        $nonce = random_bytes(\SODIUM_CRYPTO_STREAM_NONCEBYTES);
+
+        $key = $this->getKey();
+        $encryptedText = \sodium_crypto_stream_xor($message, $nonce, $key);
+        \sodium_memzero($message);
+        \sodium_memzero($key);
+
+        return $this->encode($nonce, $encryptedText);
+    }
+
+    private function getKeyPath(): ?string
+    {
+        if (null === @\is_file($this->encryptionKeyOrPath)) {
+            throw new \InvalidArgumentException('The path to the encryption key is not valid. Did you configured the framework with the content of key instead of its location?');
+        }
+
+        return $this->encryptionKeyOrPath;
+    }
+
+    private function getKey(): string
+    {
+        if (isset($this->encryptionKey)) {
+            return $this->encryptionKey;
+        }
+        // is_file returns null when the argument is not a valid string, meaning `encryptionKeyOrPath` contains the key itself
+        if (null === $isFile = @\is_file($this->encryptionKeyOrPath)) {
+            return $this->encryptionKey = $this->encryptionKeyOrPath;
+        }
+        if (true === $isFile) {
+            return $this->encryptionKey = file_get_contents($this->encryptionKeyOrPath);
+        }
+        // if `encryptionKeyOrPath` is a valid string with the exact length of a KEY, and the file does not exist, we presume it's the key itself
+        if (\SODIUM_CRYPTO_STREAM_KEYBYTES === \strlen($this->encryptionKeyOrPath)) {
+            return $this->encryptionKey = $this->encryptionKeyOrPath;
+        }
+
+        throw new EncryptionKeyNotFoundException($this->encryptionKeyOrPath);
+    }
+
+    private function encode(string $nonce, string $encryptedText): string
+    {
+        return $nonce.$encryptedText;
+    }
+
+    public function decrypt(string $encryptedText): string
+    {
+        [$nonce, $encryptedText] = $this->decode($encryptedText);
+
+        $key = $this->getKey();
+        $plainText = \sodium_crypto_stream_xor($encryptedText, $nonce, $this->getKey());
+        \sodium_memzero($key);
+
+        return $plainText;
+    }
+
+    /**
+     * @return array [$nonce, $encryptedText]
+     */
+    private function decode(string $message): array
+    {
+        if (\strlen($message) < \SODIUM_CRYPTO_STREAM_NONCEBYTES) {
+            throw new \UnexpectedValueException(sprintf('Invalid encrypted text, message should be at least %s chars long.', \SODIUM_CRYPTO_STREAM_NONCEBYTES));
+        }
+
+        $nonce = substr($message, 0, \SODIUM_CRYPTO_STREAM_NONCEBYTES);
```

Comment:
why did you use `\strlen` but not `\substr` ?

Reply:
`strlen` is "opcodes optimized", but not `substr`.

FYI: a list of method is here https://github.com/FriendsOfPHP/PHP-CS-Fixer/issues/3048#issuecomment-328810072